### ISSUE TITLE
Tags should sort numerically not alphabetically

### ISSUE THREAD

**Issue by ryandesign** [buildbot/buildbot#5424]
In the waterfall view at least, tags are sorting alphabetically. I would like them to sort numerically instead. For example, I have tags that appear in this order: 10.10 10.11 10.8 10.9. I would like them to appear in this order: 10.8 10.9 10.10 10.11.

In buildbot 0.8 we fixed this problem by using `util.naturalSort()` in several places ([one](https://github.com/macports/macports-ports/blob/0e6c74213e75ede511f329c5b670568458a7617a/devel/buildbot-0.8/files/patch-buildbot-status-web-console.py.diff), [two](https://github.com/macports/macports-ports/blob/0e6c74213e75ede511f329c5b670568458a7617a/devel/buildbot-0.8/files/patch-buildbot-status-web-grid.py.diff), [three](https://github.com/macports/macports-ports/blob/0e6c74213e75ede511f329c5b670568458a7617a/devel/buildbot-0.8/files/patch-buildbot-status-web-waterfall.py.diff)). I'm not sure where the equivalent places are in the current buildbot code. This may be one of them:

https://github.com/buildbot/buildbot/blob/494dafd8cdf4568b7451efe53d1c8c9f578f6cc3/www/waterfall_view/src/module/main.module.js#L742

I install buildbot from pypi which has precompiled frontend files, so I can't just edit that line and try it. I understand that I would need to follow the [javascript frontend environment](http://docs.buildbot.net/latest/developer/quickstart.html#create-a-javascript-frontend-environment) instructions to get something editable but that's too much for me right now.

**Comment by tardyp** [comment]
There is a naturalSort in javascript codebase:

https://github.com/buildbot/buildbot/blob/f0b3f1ef84bb6559f800d4f0001ef2cc35bd002f/www/console_view/src/module/main.module.js#L86

But it is not factorized.. So we should extract that into an angularJS filter service, so that it is available to plugins

**Comment by mojca** [comment]
@rajdeepbharati

**Comment by rajdeepbharati** [comment]
@tardyp Isn't it already available inside the console plugin? Could you give pointers on how to extract it into a filter service?

My tags are getting sorted numerically after applying this patch:

```diff
diff --git a/www/waterfall_view/src/module/main.module.js b/www/waterfall_view/src/module/main.module.js
index 93cf3addf..a98002b20 100644
--- a/www/waterfall_view/src/module/main.module.js
+++ b/www/waterfall_view/src/module/main.module.js
@@ -102,6 +102,21 @@ var WaterfallController = (function() {
                 buildidBackground: this.s.number_background_waterfall.value
             };
 
+            if (typeof Intl !== 'undefined' && Intl !== null) {
+                const collator = new Intl.Collator(undefined, {numeric: true, sensitivity: 'base'});
+                this.strcompare = collator.compare;
+            } else {
+                this.strcompare = function(a, b) {
+                    if (a < b) {
+                        return -1;
+                    }
+                    if (a === b) {
+                        return 0;
+                    }
+                    return 1;
+                };
+            }
+
             // Load data (builds and builders)
             this.all_builders = this.dataAccessor.getBuilders({order: 'name'});
             this.$scope.builders = (this.builders = this.all_builders);
@@ -739,7 +754,7 @@ var WaterfallController = (function() {
                     }
                 }
             }
-            all_tags.sort();
+            all_tags.sort(this.strcompare);
             this.$scope.all_tags = this.all_tags = all_tags;
             this.dataProcessorService.addStatus(this.builders);
```

**Comment by tardyp** [comment]
Indeed. It is crappy to have to copy paste the same code in every plugin..

There is an example of a filter here:

https://github.com/buildbot/buildbot/blob/bd27e31e6a1b5363d7f060066d6687c783a2d261/www/base/src/app/common/filters/moment/moment.filter.js#L14

It is still written as a class which is probably not the simplest way.

At this point the angularJS tutorial will help you better:
https://docs.angularjs.org/guide/filter#creating-custom-filters

Then we will be able to use your filter from any template, or we will be able to inject it as a dependency.
 
Here you can see how publicFieldFilter is just used in the JS code:

https://github.com/buildbot/buildbot/blob/bd27e31e6a1b5363d7f060066d6687c783a2d261/www/base/src/app/builders/buildrequest/buildrequest.controller.js#L8

**Comment by rajdeepbharati** [comment]
Thanks.
